### PR TITLE
PGO: update the existing benchmarks workflow to enable PGO builds

### PR DESCRIPTION
**Pull request by 1pkg**
## Motivation/summary

This PR implements changes outlined in https://github.com/elastic/apm-server/issues/13859. It updates the existing benchmarks workflow to run standalone APM Server instance that produces a relevant CPU profile for PGO, then it copies, uploads and injects the obtained CPU profile into a PR, see [example](https://github.com/elastic/apm-server/pull/14084).

## Benchmarks

The existing benchmarks results turned to be too unreliable to base PGO on. Because of the underlying dependency on ElasticSearch the difference in the throughput results could go above 10% from a workflow to workflow. The table below provides a view with the existing benchmarks results sample.

image

This all renders incremental PGO performance gains hard to observe and measure. Therefore, in this PR a new benchmark mode is introduced, which swaps ElasticSearch with a stubbed API http server (Moxy). Thus allowing us to better isolate and elevate APM Server performance component inside the benchmarks. The table below provides a view with the new isolated benchmarks results sample.

image

Using the benchmarks result sample data we can clearly observe that the results deviation for the new benchmark mode is in an order of magnitude lower in comparison to the existing ES based benchmarks. And now PGO performance improvements could be reliably observed.

The standalone APM Server benchmarks mode consists of running 3 separate EC2 instances in a VPC for apmbench, apm-server and moxy. Existing `benchmark_executor` and `standalone_apm_server` terraform modules are reused and a similar new terraform module `moxy` is created.

## Results

PGO enabled builds show 5% performance gain on average across the standalone APM Server benchmarks workflow.

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

To observe and validate the changes please refer to the indexed PGO benchmarks results.

## Related issues

#13859

### CODE REVIEW

**Comment by inge4pres** [comment]
the only doubt I have is o the CI part, not sure how to test it

**Comment by 1pkg** [comment]
> the only doubt I have is on the CI part, not sure how to test it

Github Actions are problematic to test. AFAIK, the simplest way to test them is to just run them in Github directly, which is really manual and less than ideal. I know about few projects that can help reduce the "pain":
- https://github.com/nektos/act that allows running Github Actions locally; it falls short for a complex actions like https://github.com/elastic/apm-server/actions/workflows/benchmarks.yml.
- https://github.com/sethvargo/go-githubactions that allows creating Github Actions in go; I never used it in the pass, but it looks promising. The main problem with using go-githubactions, is that it's sufficiently different from all our existing Github Actions.

**Review by inge4pres — DISMISSED** [review]
🚀 ship it!
looking forward to see it in action

**Review by marclop — COMMENTED** [review]
I didn't review in detail, just posting a half review with the top files reviewed.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml]: Agree it's useful sometimes to benchmark non "main" branch like I did in this very PR. We also can easily filter the results in the dashboard directly, so there is no concern about intermixing the results.

File: testing/benchmark/main.tf
Comment (1pkg) [reply testing/benchmark/main.tf]: Good catch, I agree 1 AZ should be good enough here, updated.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml:49]: Yes, we need `pull-requests: write` since my current idea is to raise a PR for every benchmark PGO update so we can quickly eyeball it and confirm that the source benchmarks that created it was not a failure. After we get some confidence with this approach I'd move forward with auto merging PGO.

Comment (1pkg) [reply .github/workflows/benchmarks.yml:49]: See the current workflow action [here](https://github.com/elastic/apm-server/pull/13884/files#diff-22eef62c9d7b36d02f5fd5aaada92702e22d6795c464239cb7b8fe0f26ea1e1cR195) that requires the access.

File: testing/benchmark/Makefile
Comment (marclop) [inline testing/benchmark/Makefile:134]: How come we aren't populating?
- `elasticsearch_size`
- `apm_server_zone_count`
- `elasticsearch_zone_count`

File: testing/benchmark/system-profiles/16GBx2zone.tfvars
Comment (marclop) [inline testing/benchmark/system-profiles/16GBx2zone.tfvars:23]: Wow these are pretty big and beefy. Maybe we could add a comment here. Also looking at the size, `c6i.4xlarge` seems to be a 32GB instance, not 16GB. Is that what you intended?

File: testing/benchmark/system-profiles/32GBx2zone.tfvars
Comment (marclop) [inline testing/benchmark/system-profiles/32GBx2zone.tfvars:24]: Same here, 8xlarge has 64GB of RAM.

File: testing/infra/terraform/modules/standalone_apm_server/outputs.tf
Comment (marclop) [inline testing/infra/terraform/modules/standalone_apm_server/outputs.tf:8]: I wonder if it's important at all to test with HTTPS here.

**Review by marclop — COMMENTED** [review]
mostly everything LGTM. Just a few questions we can merge soon after confirming things are as expected.

File: testing/benchmark/system-profiles/16GBx2zone.tfvars
Comment (1pkg) [reply testing/benchmark/system-profiles/16GBx2zone.tfvars]: good catch, the sizing is fixed now

File: testing/benchmark/system-profiles/32GBx2zone.tfvars
Comment (1pkg) [reply testing/benchmark/system-profiles/32GBx2zone.tfvars]: good catch, the sizing is fixed now

File: testing/benchmark/Makefile
Comment (1pkg) [reply testing/benchmark/Makefile:134]: For Moxy based benchmarks there is no ES to speak of, we can add fake value for zone count=1 to keep the consistent tagging, if we think it's a good idea.

File: testing/infra/terraform/modules/standalone_apm_server/outputs.tf
Comment (1pkg) [reply testing/infra/terraform/modules/standalone_apm_server/outputs.tf:8]: It's debatable but I think https shouldn't really matter here, since it shouldn't qualitatively change the resulting profile.

File: testing/benchmark/Makefile
Comment (marclop) [reply testing/benchmark/Makefile:134]: oh of course there's no Elasticsearch. Ignore my comment then.

**Review by marclop — DISMISSED** [review]
Thanks Kostya! Looks good now :)

**Comment by 1pkg** [comment]
Had to make some final changes to address some regression in shared modules for smoke-tests.
Now everything works as expected and aligned with the smoke tests results from the main branch.

See the workflow runs:

https://github.com/elastic/apm-server/actions/runs/11076905822
https://github.com/elastic/apm-server/actions/runs/11076854721
https://github.com/elastic/apm-server/actions/runs/11078611143

With this, this PR should be good to merge finally! Waiting for the final approvals + review from the robots team.

**Comment by v1v** [comment]
Is this automation creating cloud resources but not deleting them? I see some leftovers in 

image


that's the AWS account used for the CI - we separated the Cloud accounts from dev/ci, that's the reason you cannot access to that one

**Comment by 1pkg** [comment]
> Is this automation creating cloud resources but not deleting them? I see some leftovers in
> 
> image
> that's the AWS account used for the CI - we separated the Cloud accounts from dev/ci, that's the reason you cannot access to that one

@v1v initially there was an unrelated authentication problem in the benchmarks pipeline that caused the issue, see for example https://github.com/elastic/apm-server/actions/runs/10567550442/job/29276727544. Because I was experimenting a lot at the same time, I accidentally created multiple VPC resources that were not properly cleaned. This long got fixed since then. And now this pipeline works as expected.

Please take a closer look at the actual changes so we can merge them to the upstream, thank you.

File: .github/workflows/benchmarks.yml
Comment (v1v) [inline .github/workflows/benchmarks.yml:117]: If these env variable are only used within the context of this step, you can use the `env`


```yaml
      - name: Init terraform module
        run: make init
        env:
          TF_VAR_worker_region: ${{ env.AWS_REGION }}
          TF_VAR_run_standalone: ${{ env.RUN_STANDALONE }}
```

Otherwise, do you think you can set them in the top level env? `L54`

File: .github/workflows/benchmarks.yml
Comment (v1v) [inline .github/workflows/benchmarks.yml:128]: The condition might help to know if the step has been triggered or no easily:

```suggestion
        if: env.RUN_STANDALONE == 'true'
        run: |
          make apm-server
          make moxy
```

File: .github/workflows/benchmarks.yml
Comment (v1v) [inline .github/workflows/benchmarks.yml:155]: step native condition

```suggestion
        if: ${{ failure() && env.RUN_STANDALONE == 'true' }}
        run: make cat-apm-server-logs
```

File: .github/workflows/benchmarks.yml
Comment (v1v) [inline .github/workflows/benchmarks.yml:153]: What's the reason it's not always?

File: .github/workflows/benchmarks.yml
Comment (v1v) [inline .github/workflows/benchmarks.yml:208]: ditto.

OTOH, do you think this snippet could be added to the Makefile?

File: .github/workflows/benchmarks.yml
Comment (v1v) [inline .github/workflows/benchmarks.yml:224]: This will not trigger a build when the PR is created. Unfortunately, GitHub does not trigger builds if the actor is a `github-actions bot`. For such, there are two alternatives:

1) Use PAT or finer-grained tokens
2) Or a GitHub app to create ephemeral tokens.


`1)` is not anymore what we use, but `2)`, you can use the GitHub app as done in https://github.com/elastic/apm-server/pull/14217/files#diff-68bec17d560c218f68b50a1e84ac73026a82e64927e706cffff69501cd151970R60-R83

File: testing/infra/terraform/modules/standalone_apm_server/variables.tf
Comment (v1v) [inline testing/infra/terraform/modules/standalone_apm_server/variables.tf:60]: What's the reason for deleting these labels?

**Review by v1v — COMMENTED** [review]
I reviewed the GitHub workflow and TF variables, the other bits and pieces are not something I can say much though. I left a few comments

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml:117]: This step sets variables for terraform, I'm not 100% sure if `terraform init` is the step than binds the ENV to variables or it happens during `terraform apply` let me double check this and update it if it works in `terraform init`.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml:128]: Unfortunately, Github Actions do no support condition on env variables natively. There is a workaround to pass it as a step output, let me try to check if it works better.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml:208]: Good point, will move this to Makefile

File: testing/infra/terraform/modules/standalone_apm_server/variables.tf
Comment (1pkg) [reply testing/infra/terraform/modules/standalone_apm_server/variables.tf:60]: This module didn't use those variables, so I decided to replace it unified `tags` variable instead.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml]: Updated via steps output now. Also apparently TF binds ENVs to variables at `terraform apply`.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml]: Updated via steps outputs now.

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml]: updated

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml]: @v1v thank you for pointing to this, I updated the workflow and it uses the same plumbing to setup all necessary pieces to prepare the github config and the access token. Now the runs work as expected see [link](https://github.com/elastic/apm-server/actions/runs/11134770091/job/30943589134) and the raised [PR](https://github.com/elastic/apm-server/pull/14229) has properly signed commits.

**Comment by 1pkg** [comment]
Final results after feedback from @v1v to properly set github access token.

The PGO standalone benchmark workflow run [link](https://github.com/elastic/apm-server/actions/runs/11134770091/job/30943589134) -> resulted in the next PGO update PR [link](https://github.com/elastic/apm-server/pull/14229).

While the old benchmark against ES cloud works as expected without regression [link](https://github.com/elastic/apm-server/actions/runs/11134115634/job/30941670041).

**Review by v1v — APPROVED** [review]

File: testing/benchmark/main.tf
Comment (axw) [inline testing/benchmark/main.tf:36]: This is also in ci_tags above, should we remove it from there?

File: .github/workflows/benchmarks.yml
Comment (axw) [inline .github/workflows/benchmarks.yml:210]: Why do we need this? Can't we add `pull_requests: write` to the workflow's permissions?

Is it because we are creating a PR that triggers another workflow?

File: .github/workflows/benchmarks.yml
Comment (axw) [inline .github/workflows/benchmarks.yml:235]: I wonder if instead of creating a new PR on every benchmark run, could we just push a commit to the branch?

**Review by axw — DISMISSED** [review]
Nice work - thank you for all the cleanups along the way!

File: .github/workflows/benchmarks.yml
Comment (v1v) [reply .github/workflows/benchmarks.yml:210]: > Is it because we are creating a PR that triggers another workflow?

Correct, as per the [documentation](https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow), 'When you use the repository's GITHUB_TOKEN to perform tasks, events triggered by the GITHUB_TOKEN, with the exception of workflow_dispatch and repository_dispatch, will not create a new workflow run.'

https://github.com/elastic/apm-server/pull/13884#discussion_r1783310922 asked to change it

File: testing/benchmark/Makefile
Comment (v1v) [reply testing/benchmark/Makefile:193]: That's me who I asked for moving the logic to a Makefile in https://github.com/elastic/apm-server/pull/13884#discussion_r1783309017

File: testing/benchmark/outputs.tf
Comment (1pkg) [reply testing/benchmark/outputs.tf]: good catch, marked by a mistake, updated

File: testing/benchmark/Makefile
Comment (1pkg) [reply testing/benchmark/Makefile]: Moved to .ci/scripts/ now

File: testing/benchmark/main.tf
Comment (1pkg) [reply testing/benchmark/main.tf]: yes, it's a leftover, updated

File: .github/workflows/benchmarks.yml
Comment (1pkg) [reply .github/workflows/benchmarks.yml]: I think it's slightly risky to enable auto pushes to main branch right away, I'd prefer to start with more controlled PR based approach so we develop the confidence that this pipeline works well. Afterwards we can simplify and enable the direct merge.

Comment (1pkg) [reply .github/workflows/benchmarks.yml]: I also added a small update to the push-pgo-pr script so it enables auto merging too for PRs. This way we will only need to give it 1 approval and the pipeline tests need to pass.

Comment (axw) [reply .github/workflows/benchmarks.yml]: Fair enough.

**Review by axw — APPROVED** [review]
LGTM, thank you!

**Pull request merged**
